d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -114,91 +115,117 @@ public WarmUpLoadBalancer(LoadBalancerWithFacilities balancer, WarmUpService ser
   @Override
   public void start(Callback<None> callback) {
     LOG.info("{} enabled", _printName);
-    _loadBalancer.start(new Callback<None>() {
-      @Override
-      public void onError(Throwable e) {
-        callback.onError(e);
-      }
 
+    Callback<None> prepareWarmUpCallback = new Callback<None>() {
       @Override
-      public void onSuccess(None result) {
-        _allStartTime = SystemClock.instance().currentTimeMillis();
-
-        Future<?> prepareTaskFuture = _executorService.submit(() -> prepareWarmUp());
-        try
-        {
-          prepareTaskFuture.get(_warmUpTimeoutSeconds, TimeUnit.SECONDS);
-        }
-        catch (TimeoutException e)
+      public void onError(Throwable e) {
+        if (e instanceof TimeoutException)
         {
           LOG.info("{} hit timeout: {}s. The WarmUp will continue in background", _printName, _warmUpTimeoutSeconds);
           callback.onSuccess(None.none());
         }
-        catch (Exception e)
+        else
         {
           LOG.error("{} failed to fetch dual read mode, continuing warmup.", _printName, e);
         }
```

Comment:
Should call `callback.onError(e)` here?

Reply:
not really, since this is just the preparation phase (fetching the dual read mode), the actual warmup phase can still warm up more data if it haven't hit timeout before letting the startup complete. (In fact, if startup callback's onError is called, the startup will fail, which shouldn't be caused by warmup since warmup is not required.)